### PR TITLE
add logger preamble and misfire_grace_time

### PR DESCRIPTION
**Pull request by vict0rsch**
Addressing #238 

```python
In [5]: e = EmissionsTracker(logger_preamble=" [Node1]")
[codecarbon INFO @ 15:52:23] [Node1] Tracking Intel CPU via Power Gadget
```

and 

```
        self._scheduler.add_job(
            self._measure_power,
            "interval",
            seconds=self._measure_power_secs,
            max_instances=1,
            misfire_grace_time=self._misfire_grace_time,
        )
```

### CODE REVIEW

**Comment by stas00** [comment]
So I did several experiments:

- measure_power_secs=30 still warns, but much less often:
```
Run time of job "BaseEmissionsTracker._measure_power (trigger: interval[0:00:30], next run at: 2021-08-28 02:00:13 CEST)" was missed by 0:00:03.089666
```

- `max_instances=3` seems to not generate warnings, at least in the few experiments I have run - but this one requires hacking the source code, so can we make it an argument to the base Tracker class?

**Comment by stas00** [comment]
oh, just noticed another one:
```
[codecarbon DEBUG @ 02:22:45] r007 Exception occurred while instantiating IntelPowerGadget : Platform not supported by Intel Power Gadget
Traceback (most recent call last):
  File "/gpfsssd/worksf/projects/rech/six/ura81os/code/codecarbon/codecarbon/core/cpu.py", line 27, in is_powergadget_available
    IntelPowerGadget()
  File "/gpfsssd/worksf/projects/rech/six/ura81os/code/codecarbon/codecarbon/core/cpu.py", line 66, in __init__
    self._setup_cli()
  File "/gpfsssd/worksf/projects/rech/six/ura81os/code/codecarbon/codecarbon/core/cpu.py", line 93, in _setup_cli
    raise SystemError("Platform not supported by Intel Power Gadget")
SystemError: Platform not supported by Intel Power Gadget
```
early on. Not sure if it was always there. I think it only emerged when debug was turned on, so probably it's ok.

**Comment by stas00** [comment]
Not sure if relativism kicks in when one puts the magnifying glass on, but I can't get any runtime warnings about misfire if I enable debug - perhaps it slows everything down long enough for everything to succeed.

I've attached the log file (just the carbon logs) with the debug dump if you will find any good use for it.

[carbon.log.txt](https://github.com/mlco2/codecarbon/files/7069882/carbon.log.txt)

**Comment by benoit-cty** [comment]
For IntelPowerGadget, yes it's due to debug. It just mean you are not in Windows, we have to made this look less scary.

Your log looks good. If your GPU usage is quite constant you may increase measure_power_secs to took measures every 300 s instead of 15s.

But all your machines measures are in a window of 2 seconds so it won't improve anything, just reduce log size.

**Comment by stas00** [comment]
Great, will use a much longer `measure_power_secs` - I suppose on `stop` it will still take a full measure even if it's before  `measure_power_secs`, correct?

Also, a slightly unrelated question - do we need to do any exception handling on behalf of CC? It's an optional feature and must absolutely not crash the main training should something go wrong.

Thank you!

File: codecarbon/external/logger.py [hunk 1]
```diff
@@ -1,18 +1,61 @@
 import logging
 import os
+from typing import Optional
+
+
+def set_logger_format(custom_preamble: Optional[str] = ""):
+    logger = logging.getLogger("codecarbon")
+    format = f"[%(name)s %(levelname)s @ %(asctime)s]{custom_preamble} %(message)s"
```
Comment (benoit-cty) [inline codecarbon/external/logger.py:8]: Space or brackets could be added here for custom_preamble..

**Review by benoit-cty — APPROVED** [review]

**Comment by benoit-cty** [comment]
Yes, on stop or flush we compute power based on the time between last measure.
The only drawback is that you probably stop your training before stopping CC, so CC will not see a realistic power consumption on stop : for CPU we have a real Joules consumption but for GPU we only have the instant power consumption at the exact time of measurement. So the measure_power_secs must not be too big as our measurement is not rocket science...

For crash we have `@suppress(Exception)` to avoid that on start / stop / flush, but @vict0rsch maybe we have to add it on `_measure_power` or it's already handled by the scheduler ? Because I saw we could `raise ValueError` in `_measure_power`

**Comment by stas00** [comment]
We call `cc.stop` at the very end of the program, so after training has completed. But there are some cases where it `sys.exit()` on a different path. can probably try and catch those cases too. But since it saves the checkpoint after finishing the training, this is probably just as good.

File: codecarbon/external/logger.py (hunk 1 above)
Comment (vict0rsch) [reply codecarbon/external/logger.py]: sold, adding brackets

**Comment by vict0rsch** [comment]
Summary of latests commits:

* `max_instances` argument for the scheduler's jobs
* preamble inside brackets (`[codecarbon INFO @ 02:22:45][preamble] ...`)
* no more `ValueError` in `_measure_power()`, rather a `logger.error`
* clean up the PowerGadget & RAPL `debug` logs so they look more normal

**Review by benoit-cty — APPROVED** [review]

**Comment by stas00** [comment]
could this be merged, please?

**Comment by stas00** [comment]
Thank you, @vict0rsch!

**Pull request merged**